editoast/core_task/src/envs/simulation/request.rs

```diff
@@ -0,0 +1,426 @@
+use std::hash::Hash;
+use std::sync::Arc;
+
+use core_client::AsCoreRequest as _;
+use core_client::CoreClient;
+
+use crate::CoreEnv;
+use crate::Task;
+use crate::envs::pathfinding;
+
+use super::SimulationKey;
+use super::SimulationWaypointSchedule;
+
+pub(super) fn build_request(
+    core_env: &CoreEnv,
+    electrical_profiles_set_id: Option<u64>,
+    SimulationKey(sim_consist, params, pathfinding::Input(pf_consist, constraints)): &SimulationKey,
+    core_client::pathfinding::PathfindingResultSuccess {
+        path,
+        path_item_positions,
+        ..
+    }: core_client::pathfinding::PathfindingResultSuccess,
+) -> Result<core_client::simulation::Request, ()> {
+    if constraints.path_items.len() != path_item_positions.len() {
+        tracing::error!(
+            ?sim_consist,
+            ?pf_consist,
+            sim_params = ?params,
+            pf_constraints = ?constraints,
+            waypoint_count = constraints.path_items.len(),
+            computed_path_item_count = path_item_positions.len(),
+            "Core path does not respect pathfinding constraints. Please open a bug report."
+        );
+        return Err(());
+    }
+    // Thanks to SimulationTrain, all indexing from now on should be safe.
+    // If any indexing below goes wrong, it's a bug.
+
+    let mut power_restrictions = Vec::new();
+    for (range, restriction) in params.power_restrictions.iter() {
+        let from = path_item_positions[range.start];
+        let to = path_item_positions[range.end];
+        power_restrictions.push(core_client::simulation::SimulationPowerRestrictionItem {
+            from,
+            to,
+            value: restriction.clone(),
+        });
+    }
+
+    let mut margin_boundaries = Vec::<u64>::new();
+    let mut margin_values = Vec::new();
+    for (range, margin) in params.margins.iter() {
+        let from = path_item_positions[range.start];
+        let to = path_item_positions[range.end];
+        if let Some(prev) = margin_boundaries.last() {
+            debug_assert_eq!(*prev, from);
+        } else {
+            margin_boundaries.push(from);
+        }
+        margin_boundaries.push(to);
+        margin_values.push(*margin);
+    }
+    debug_assert!(
+        margin_boundaries.is_empty() && margin_values.is_empty()
+            || margin_boundaries.len() == margin_values.len() + 1
+    );
+
+    let mut schedule = Vec::new();
+    for (i, point) in params.schedule.iter().enumerate() {
+        if let SimulationWaypointSchedule::Stop {
+            arrival_at,
+            stop_for,
+            reception_signal,
+        } = point
+        {
+            let position = path_item_positions[i];
+            schedule.push(core_client::simulation::SimulationScheduleItem {
+                path_offset: position,
+                arrival: *arrival_at,
+                stop_for: *stop_for,
+                reception_signal: *reception_signal,
+            });
+        }
+    }
```

Comment:
nitpick: The `if` is a **filter**, and therefore, perhaps it would make sense to express it as such. On top of that, it would avoid the `mut` 😉 


```suggestion
    let schedule = params.schedule.iter()
        .enumerate()
        .filter_map(|(i, point)| {
            if let SimulationWaypointSchedule::Stop {
                arrival_at,
                stop_for,
                reception_signal,
            } = point
            {
                let position = path_item_positions[i];
                Some(core_client::simulation::SimulationScheduleItem {
                    path_offset: position,
                    arrival: *arrival_at,
                    stop_for: *stop_for,
                    reception_signal: *reception_signal,
                })
            } else {
                None
            }
        )
        .collect_vec();
```

Reply:
I'm fine with functional style (as you'd expect) but since this function is written in imperative style I thought it'd be best not to mix and match